File: Jonte/day25/handshake.cpp

#include <algorithm>
#include <chrono>
#include <iostream>
#include <vector>

namespace cs = std::chrono;

constexpr int subject_number = 7;
constexpr int denominator = 20201227;

int compute_key(size_t loop_size) {
  int val = 1;

  for (size_t i = 0; i < loop_size; ++i) {
    val *= subject_number;
    val = val % denominator;
  }

  return val;
}

int get_next(int val) {
  int next_val = val * subject_number;
  return next_val % denominator;
}

int main() {
  size_t max_size = 1e8;

  std::vector<int> values(max_size);

  auto t0 = cs::high_resolution_clock::now();

  values[0] = 1;
  for (size_t i = 1; i < max_size; ++i) {
    values[i] = get_next(values[i - 1]);
  }

  auto t1 = cs::high_resolution_clock::now();

  std::cout << "computed values for " << max_size << " loop sizes in "
            << cs::duration_cast<cs::milliseconds>(t1 - t0).count() << " ms\n";            

  int card_public_key = 2959251;
  int door_public_key = 4542595;

  auto it_card = std::find(values.begin(), values.end(), card_public_key);
  if (it_card != values.end()) {
    std::cout << "card loop size: " << std::distance(values.begin(), it_card)
              << "\n";
  } else {
    std::cout << "unable to find card loop size\n";
  }

  auto it_door = std::find(values.begin(), values.end(), door_public_key);
  if (it_door != values.end()) {
    std::cout << "door loop size: " << std::distance(values.begin(), it_door)
              << "\n";
  } else {
    std::cout << "unable to find door loop size\n";
  }
}

File: Jonte/day25/handshake.py

# AoC 2020 Day 25

def compute_key(subject_number, loop_size):
    val = 1
    denominator = 20201227

    for _ in range(loop_size):
        val = val * subject_number
        val = val % denominator
    
    return val

subject_number = 7
denominator = 20201227

def next_value(val):
    next_val = val * subject_number
    return next_val % denominator


# values for all loop sizes up to 1e8

max_size = 100000000
values = [1] * max_size

for i in range(1, max_size):
    values[i] = next_value(values[i - 1])

card_public_key = 2959251
door_public_key = 4542595

loop_size_card = next(i for i, v in enumerate(values) if v == card_public_key)
loop_size_door = next(i for i, v in enumerate(values) if v == door_public_key)

print(loop_size_card, loop_size_door)

encryption_key = compute_key(door_public_key, loop_size_card)
print(encryption_key)


